### ISSUE TITLE
Issue with cyclic TypedDict definitions

### ISSUE THREAD

**Issue by bijij** [microsoft/pyright#2886]
**Describe the bug**
Pyright adds unexpected types to a union which cases a typeddict to error as not subscriptable.

**To Reproduce**
See the code below,
Pyright will output the following
```
test2.py
  test2.py:31:17 - information: Type of "a" is "BarA | BarB"
  test2.py:33:17 - information: Type of "options" is "Any | List[Type[BarA] | Type[BarB]]" -- where did the Type come from?
  test2.py:35:21 - information: Type of "option" is "Any | Type[BarA] | Type[BarB]"
  test2.py:36:21 - error: Object of type "AllBar" is not subscriptable (reportGeneralTypeIssues) -- ???
  test2.py:36:21 - information: Type of "option["type"]" is "Any | Unknown"
  ```
  
  interestingly if `Foo` which is not being used anywhere in the `foo` function is deleted or moved after the definition of `BarB` the code will type check, outputting something like

```
test3.py
  test3.py:32:17 - information: Type of "a" is "BarA | BarB"
  test3.py:34:17 - information: Type of "options" is "Any | List[BarA | BarB]"
  test3.py:36:21 - information: Type of "option" is "Any | BarA | BarB"
  test3.py:37:21 - information: Type of "option["type"]" is "Any | int"
  ```

**Expected behavior**
No matter the order of definitions, pyright should probably be outputting what is shown with `test3.py`, (preferably the any should be dropped but I understand why it is there, do you have any suggestions on how to avoid it?)

**Screenshots or Code**
I've tried to make this as concise as possible, the code which I originally found this in was considerably more complex, this is the smallest form I could find.
```python
from __future__ import annotations

from typing import Union, TypedDict, List


class _FooOptional(TypedDict, total=False):
    options: List[AllBar]
    type: int


class Foo(_FooOptional):
    id: int
    name: str


class BarA(TypedDict):
    type: int

class BarB(TypedDict):
    options: List[AllBar]
    type: int


AllBar = Union[
    BarA,
    BarB
]


def foo(a: AllBar):
    reveal_type(a)
    options = a.get("options", [])
    reveal_type(options)
    for option in options:
        reveal_type(option)
        reveal_type(option["type"])

```

**VS Code extension or command-line**
pyright 1.1.210

**Comment by erictraut** [comment]
Thanks for the bug report. There was a bug in the code that evaluates index expressions like `option["type"]`. It wasn't properly dealing with a recursive type alias like `AllBar`. This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.211, which I just published. It will also be included in the next release of pylance.
